source/2_platform/rex_directx/src/directx_renderer.cpp

```diff
@@ -73,76 +238,1816 @@ namespace rex
+            return false;
+          }
+
+          if(FAILED(g_ctx.device->CreateCommandList(0, D3D12_COMMAND_LIST_TYPE_DIRECT, g_ctx.command_allocator.Get(), nullptr, IID_PPV_ARGS(g_ctx.command_list.GetAddressOf()))))
+          {
+            REX_ERROR(LogDirectX, "Failed to create command list");
+            return false;
+          }
+
+          // Start off in a closed state. This is because the first time we
+          // refer to the command list we will Reset it, and it needs to be closed
+          // before calling Reset.
+          if(close_command_list() == false)
+          {
+            REX_ERROR(LogDirectX, "Failed to close command list");
+            return false;
+          }
+
+          return true;
+        }
+
+        //-------------------------------------------------------------------------
+        bool create_swapchain_object(dxgi::Factory* factory, const OutputWindowUserData& userData)
+        {
+          g_ctx.swapchain.Reset();
+
+          DXGI_SWAP_CHAIN_DESC sd;
+          sd.BufferDesc.Width                   = userData.window_width;
+          sd.BufferDesc.Height                  = userData.window_height;
+          sd.BufferDesc.RefreshRate.Numerator   = userData.refresh_rate;
+          sd.BufferDesc.RefreshRate.Denominator = 1;
+          sd.BufferDesc.Format                  = g_ctx.back_buffer_format;
+          sd.BufferDesc.ScanlineOrdering        = DXGI_MODE_SCANLINE_ORDER_UNSPECIFIED;
+          sd.BufferDesc.Scaling                 = DXGI_MODE_SCALING_UNSPECIFIED;
+          sd.SampleDesc.Count                   = g_ctx.msaa_state ? 4 : 1;
+          sd.SampleDesc.Quality                 = g_ctx.msaa_state ? g_ctx.msaa_quality - 1 : 0;
+          sd.BufferUsage                        = DXGI_USAGE_RENDER_TARGET_OUTPUT;
+          sd.BufferCount                        = s_swapchain_buffer_count;
+          sd.OutputWindow                       = (HWND)userData.primary_display_handle;
+          sd.Windowed                           = userData.windowed;
+          sd.SwapEffect                         = DXGI_SWAP_EFFECT_FLIP_DISCARD;
+          sd.Flags                              = DXGI_SWAP_CHAIN_FLAG_ALLOW_MODE_SWITCH;
+
+          // Note: swap chain uses queue to perform flush.
+          rex::wrl::com_ptr<IDXGIFactory> dxgi_factory = factory->as<IDXGIFactory>();
+          if(FAILED(dxgi_factory->CreateSwapChain(g_ctx.command_queue.Get(), &sd, g_ctx.swapchain.GetAddressOf())))
+          {
+            REX_ERROR(LogDirectX, "Failed to create swap chain");
+            return false;
+          }
+
+          directx::set_debug_name_for(g_ctx.swapchain.Get(), "SwapChain");
+
+          return true;
+        }
+
+        //-------------------------------------------------------------------------
+        wrl::com_ptr<ID3D12Resource> create_default_buffer(ID3D12Device* device, ID3D12GraphicsCommandList* cmdList, const void* initData, UINT64 byteSize, wrl::com_ptr<ID3D12Resource>& uploadBuffer)
```

Comment:
uploadBuffer is output param, which is not clear on callsite, please use rsl::Out if you want to do this or find another way to indicate this

Reply:
I addressed this but rsl::optional does not compile.
So I used std this needs to be converted to RSL